src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+            }
+
+            // A negated discard node represents an always false pattern
+            private bool IsNegatedDiscard(BoundPattern pattern, bool synthesized = false)
+            {
+                if (synthesized)
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern { WasCompilerGenerated: true } };
+                }
+                else
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern };
+                }
+            }
+
+            public override BoundNode? VisitBinaryPattern(BoundBinaryPattern node)
+            {
+                var resultLeft = (BoundPattern)Visit(node.Left);
+                var resultRight = (BoundPattern)Visit(node.Right);
+                var result = node.WithLeft(resultLeft).WithRight(resultRight);
+                return _negated ? result.WithDisjunction(!node.Disjunction) : result;
+            }
+
+            public override BoundNode? VisitNegatedPattern(BoundNegatedPattern node)
+            {
+                if (_negated)
+                {
+                    return Instance.Visit(node.Negated);
+                }
+
+                return NegatedInstance.Visit(node.Negated);
+            }
+
+            public BoundPattern NegateIfNeeded(BoundPattern node)
+            {
+                if (_negated)
+                {
+                    var result = new BoundNegatedPattern(node.Syntax, node, node.InputType, narrowedType: node.InputType);
+                    if (node.WasCompilerGenerated)
+                    {
+                        result.MakeCompilerGenerated();
+                    }
+                    return result;
+                }
+
+                return node;
+            }
+
+            public override BoundNode? VisitTypePattern(BoundTypePattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+            {
+                BoundPattern result;
+                if (node.IsVar)
+                {
+                    result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                }
+                else
+                {
+                    if (node.InputType.Equals(node.DeclaredType.Type)) // TODO2 what kind of type comparison?
+                    {
+                        result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                    }
+                    else
+                    {
+                        result = node;
```

Comment:
>result = node; [](http://example.com/codeflow?start=24&length=14)

I tried to use ```result = MakeDiscardPattern(node).MakeCompilerGenerated();``` here instead, and that made an observable difference only in ```RedundantPattern_DisallowDesignatorsUnderNotAndOr``` and ```RedundantPattern_LeftBranches``` for error scenarios, for which I would suggest to not run redundancy analysis at all. I think we need to come up with a test scenario that illustrates a real benefit from having the enclosing `if`.